### PR TITLE
[feature-wip](resouce-group)  Supports memory soft isolation of resource group

### PR DESCRIPTION
**Pull request by luozenglin**
# Proposed changes

Issue Number: close #xxx

## Problem summary

```
create resource groups name properties(
    'enable_memory_overcommit' = 'true' // whether to enable memory soft isolation
)
```

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by luozenglin** [comment]
run buildall

**Comment by luozenglin** [comment]
run buildall

**Comment by luozenglin** [comment]
run buildall

File: be/src/util/mem_info.cpp [hunk 1]
```diff
@@ -144,6 +146,8 @@ bool MemInfo::process_minor_gc() {
     // TODO add freed_mem
     SegmentLoader::instance()->prune();
 
+    freed_mem += tg_soft_memory_limit_gc(_s_process_minor_gc_size - freed_mem);
+
```
Comment (xinyiZzz) [inline be/src/util/mem_info.cpp:150]: add
```
if (freed_mem > _s_process_minor_gc_size) {
    return true;
}
```

File: be/src/util/mem_info.cpp [hunk 2]
```diff
@@ -187,6 +191,8 @@ bool MemInfo::process_full_gc() {
         }
     }
 
+    freed_mem += tg_soft_memory_limit_gc(_s_process_minor_gc_size - freed_mem);
+
```
Comment (xinyiZzz) [inline be/src/util/mem_info.cpp:195]: same add
if (freed_mem > _s_process_minor_gc_size) {
    return true;
}

File: be/src/util/mem_info.cpp [hunk 3]
```diff
@@ -144,6 +146,8 @@ bool MemInfo::process_minor_gc() {
     // TODO add freed_mem
     SegmentLoader::instance()->prune();
 
+    freed_mem += tg_soft_memory_limit_gc(_s_process_minor_gc_size - freed_mem);
```
Comment (xinyiZzz) [inline be/src/util/mem_info.cpp:149]: Add more detailed comments for `process_minor_gc` and `process_full_gc`. GC order in each case.

Comment (xinyiZzz) [inline be/src/util/mem_info.cpp:149]: 现在memory exceed limit的行为是这样么，辛苦确认下：
group－A没有overcommit，只有一个query－a，query—a已经overcommit。
group－B已经overcommit，其中query－b已经overcommit，其他query都没有overcommit。

进程内存超限时 minor gc，
会先cancel query－b，再按照内存从大到小cancel group－B的其他query，直到group－B不再overcommit，然后再cancel query－a

File: be/src/runtime/task_group/task_group.cpp [hunk 4]
```diff
@@ -132,17 +148,23 @@ void TaskGroup::remove_mem_tracker_limiter(std::shared_ptr<MemTrackerLimiter> me
...
-        memory_limit = _memory_limit;
-    }
-    return MemTrackerLimiter::tg_memory_limit_gc(_id, name, memory_limit,
+    const auto& [memory_limit, name] = _memory_limit_and_name();
+    auto used = memory_used();
+    return MemTrackerLimiter::tg_memory_limit_gc(used - memory_limit, used, _id, name, memory_limit,
                                                  _mem_tracker_limiter_pool);
 }
 
+int64_t TaskGroup::soft_memory_limit_gc(int64_t request_free_memory, int64_t used_memory) {
```
Comment (xinyiZzz) [inline be/src/runtime/task_group/task_group.cpp:157]: Add comment, the difference between `soft_memory_limit_gc` and `memory_limit_gc` is that `soft_memory_limit_gc` request_free_memory <= used - memory_limit

maybe can change name

File: be/src/util/mem_info.cpp [hunk 5]
```diff
@@ -215,6 +221,52 @@ bool MemInfo::process_full_gc() {
...
+    for (const auto& task_group : task_groups) {
+        auto used_memory = task_group->memory_used();
+        auto exceeded = used_memory - task_group->memory_limit();
+        auto exceeded_memory = exceeded > 0 ? exceeded : 0;
+        total_exceeded_memory += exceeded_memory;
+        used_memorys.emplace_back(used_memory);
+        exceeded_memorys.emplace_back(exceeded_memory);
+    }
+
+    int64_t total_free_memory = 0;
```
Comment (wangbo) [inline be/src/util/mem_info.cpp:259]: Need a todo here: support mem priority; it's used for release overcommit memory

File: be/src/util/mem_info.cpp [hunk 6]
```diff
@@ -215,6 +221,52 @@ bool MemInfo::process_full_gc() {
...
+    for (int i = 0; i < task_groups.size(); ++i) {
+        // Use resource group exceeded memory as a weight
+        int64_t tg_need_free_memory = static_cast<double>(exceeded_memorys[i]) /
```
[the remark this answers is not in the document]
Comment (wangbo) [reply be/src/util/mem_info.cpp:262]: or maybe:
if total_exceeded_memory > request_free_memory
  this means just release some groups is ok, so release TopN overcommit memory group;
if total_exceeded_memory <= request_free_memory
   need release all groups overcommit memory.

File: be/src/runtime/memory/mem_tracker_limiter.cpp [hunk 7]
```diff
@@ -487,21 +487,14 @@ int64_t MemTrackerLimiter::free_top_overcommit_query(
 }
 
 int64_t MemTrackerLimiter::tg_memory_limit_gc(
```
Comment (wangbo) [inline be/src/runtime/memory/mem_tracker_limiter.cpp:489]: tg_hard_memory_limit_gc
-> memory_limit_gc
-> tg_memory_limit_gc
But free_top_memory_query is also called  in this method.
It seems that we just need to free_top_overcommit_query here?

**Comment by luozenglin** [comment]
run buildall

**Review by wangbo — APPROVED** [review]
LGTM

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Pull request merged**